File: typescript/prototype/message-box.ts
import lodash from 'lodash';
import { Product } from './product';

export class MessageBox implements Product {
  constructor(private decochar: string) {}

  use(str: string): void {
    const length: number = encodeURIComponent(str).replace(/%../g,"x").length;
    for(let i = 0; i < length + 4; i++) {
      process.stdout.write(this.decochar);
    }
    console.log('');
    console.log(`${this.decochar} ${str} ${this.decochar}`);
    for(let i = 0; i < length + 4; i++) {
      process.stdout.write(this.decochar);
    }
    console.log('');
  }

  createClone(): Product {
    return lodash.cloneDeep(this);
  }
}

File: kotlin/prototype/src/MessageBox.kt
package prototype

class MessageBox(private val decoChar: Char) : Product() {
    override fun use(str: String) {
        val length: Int = str.toByteArray().size
        for (i in 0..(length + 3)) {
            print(decoChar)
        }
        println()
        println("$decoChar $str $decoChar")
        for (i in 0..(length + 3)) {
            print(decoChar)
        }
        println()
    }

    override fun createClone(): Product {
        try {
            return clone() as Product
        } catch (e : CloneNotSupportedException) {
            throw e
        }
    }
}